src/lightning/pytorch/callbacks/batch_size_finder.py

```diff
@@ -128,24 +128,25 @@ def __init__(
     def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: Optional[str] = None) -> None:
         if trainer._accelerator_connector.is_distributed:
             raise MisconfigurationException("The Batch size finder is not supported with distributed strategies.")
-
-        running_stage = trainer.state.stage
-        assert running_stage is not None
-        dl_source = getattr(trainer._data_connector, f"_{running_stage.dataloader_prefix}_dataloader_source")
-
         # TODO: check if this can be enabled (#4040)
-        if not trainer._data_connector._train_dataloader_source.is_module():
+        if not trainer.fit_loop._data_source.is_module():
             raise MisconfigurationException(
                 "The Batch size finder cannot be used with dataloaders passed directly to `.fit()`. Please disable"
                 " the feature or incorporate the dataloader into your LightningModule or LightningDataModule."
             )
 
         # TODO: Add support for multiple eval dataloader
         if stage != "fit":
-            dataloaders = dl_source.dataloader()
-            if isinstance(dataloaders, list) and len(dataloaders) > 1:
+            loop = trainer._active_loop
+            assert loop is not None
```

Comment:
nit: is this just for mypy? If yes, the typing.cast solution might be better as it is a noop at runtime.

Reply:
The active loop can actually be None with ddp spawn. This is because the stage might not be set ([example](https://github.com/Lightning-AI/lightning/blob/7f92d5c9d4f309b036b46957d32771c882c931ef/src/lightning/pytorch/trainer/trainer.py#L877-L879)). We would need to include it in the queue to avoid these None checks.